ultimatepython/classes/inheritance.py

```diff
@@ -0,0 +1,134 @@
+        return f"{self} is driving at {rate_in_mph} MPH"
+
+
+class Car(Vehicle):
+    """Definition of a car.
+
+    We inherit from the vehicle class to reuse the code and data that
+    we have already defined. In addition, we add a new attribute called
+    `wheels` to the car class. This is an example of extending the
+    functionality of a parent class. In __init__, we call the parent
+    constructor with the `super` function. This is a way to invoke the
+    parent constructor without having to explicitly name the parent
+    class. We also override the method `__repr__` to have a different
+    output than the vehicle.
+    """
+
+    def __init__(self, make, model, year, miles):
+        """Construct a car with make, model, year, miles, and wheels."""
+        super().__init__(make, model, year, miles)
+        self.wheels = 4
+
+    def __repr__(self):
+        """Return the formal representation of a car."""
+        return f"<Car make={self.make} model={self.model} year={self.year} wheels={self.wheels}>"
+
+
+class Truck(Vehicle):
+    """Definition of a truck.
+
+    We inherit from vehicle just like we did with the car class. In this case we
+    will also override the method `drive` to have a different output than the
+    car and the vehicle.
+    """
+
+    def __init__(self, make, model, year, miles):
+        """Construct a truck with make, model, year, miles, and wheels."""
+        super().__init__(make, model, year, miles)
+        self.wheels = 6
+
+    def __repr__(self):
+        """Return the formal representation of a truck."""
+        return f"<Truck make={self.make} model={self.model} year={self.year} wheels={self.wheels}>"
+
+    def drive(self, rate_in_mph):
+        """Drive a truck at a certain rate in MPH."""
+        return f"{self} is driving a truck at {rate_in_mph} MPH"
+
+
+def main():
+    # Create a vehicle with the provided class constructor
+    vehicle = Vehicle("Mistery Machine", "Van", 1969, 100000.0)
+
+    # Formal representation
+    assert repr(vehicle) == "<Vehicle make=Mistery Machine model=Van year=1969>"
+
+    # Informal representation
+    assert str(vehicle) == "Mistery Machine Van (1969)"
+
+    # Call a method on the class constructor
+    assert vehicle.drive(50) == "Mistery Machine Van (1969) is driving at 50 MPH"
+
+    # Check the type of the method drive
+    assert ismethod(vehicle.drive) and not isfunction(vehicle.drive)
+
+    # Now we create a car with the provided class constructor
+    car = Car("DeLorean", "DMC-12", 1982, 220000.0)
+
+    # The informal representation is similar to the vehicle
+    assert str(car) == "DeLorean DMC-12 (1982)"
+
+    # But the formal representation is different because we included
+    # the wheels attribute
+    assert repr(car) == "<Car make=DeLorean model=DMC-12 year=1982 wheels=4>"
+
+    # And we can check the type of the method drive like we did with
+    # the vehicle
+    assert ismethod(car.drive) and not isfunction(car.drive)
```

Comment:
Should we add an assertion for `car.drive` as well for completeness?

Reply:
Sorry, I should have been more clear. `car.drive(50)` should follow similar output as `vehicle.drive(50)`. I was wondering if you wanted to elaborate on that. And maybe show the inheritance chain upwards. It may even be a sneak peak into the `mro.py` lesson!